service/history/workflow/update/update.go

```diff
@@ -351,6 +356,31 @@ func (u *Update) onResponseMsg(
 	return nil
 }
 
+func (u *Update) cancel(
+	_ context.Context,
+	eventStore EventStore,
+) {
+	const waitAccepted = stateAdmitted | stateProvisionallyRequested | stateRequested
+	const waitCompleted = stateAccepted | stateProvisionallyAccepted
+
+	prevState := u.setState(stateProvisionallyCompleted)
+	eventStore.OnAfterCommit(func(context.Context) {
+		if prevState.Matches(stateSet(waitAccepted)) {
+			u.request = nil
+			u.setState(stateCompleted)
+			u.accepted.(*future.FutureImpl[*failurepb.Failure]).Set(nil, CancelErr)
+			u.outcome.(*future.FutureImpl[*updatepb.Outcome]).Set(nil, CancelErr)
+			u.onComplete()
+		} else if prevState.Matches(stateSet(waitCompleted)) {
+			u.setState(stateCompleted)
+			u.outcome.(*future.FutureImpl[*updatepb.Outcome]).Set(nil, CancelErr)
+			u.onComplete()
```

Comment:
This doesn't deadlock? The Registry is holding an slock and calling cancel, cancel upcalls to the remover func which takes an xlock. I must be missing something but I don't see how this could work.

Reply:
Oh, it depends on the effects.Controller being passed. effects.Immediate would deadlock.